android/libraries/rib-coroutines/src/test/kotlin/com/uber/rib/core/CoroutineWorkerTest.kt

```diff
@@ -0,0 +1,232 @@
+  @Test
+  internal fun onBindingOnDifferentThread_dispatchesToScopeThread() = runTest {
+    val callingContext = newSingleThreadContext("Calling context")
+    val executionContext = newSingleThreadContext("Execution context")
+    executionContext.use { execCtx ->
+      callingContext.use { callingCtx ->
+        withContext(callingCtx) {
+          val handle = (this@runTest + execCtx).bind(worker)
+          handle.join()
+          assertThat(worker.onStartThread!!.name).startsWith("Execution context")
+          handle.unbind().join()
+          assertThat(worker.onStopThread!!.name).startsWith("Execution context")
+        }
+      }
+    }
+  }
+
+  @Test
+  internal fun onBindingOnCorrectContext_doNotPayForDispatch() = runTest {
+    val dispatcher = ImmediateDispatcher(testScheduler)
+    dispatcher {
+      dispatcher.setThreadId()
+      assertThat(dispatcher.dispatchCount).isEqualTo(1)
+      val handle = bind(worker)
+      assertThat(dispatcher.dispatchCount).isEqualTo(1) // no new dispatch done
+      assertThat(worker.onStartStarted).isTrue() // started undispatched
+      assertThat(worker.onStartThread!!.id).isEqualTo(Thread.currentThread().id)
+      // run delay on onStart
+      advanceTimeBy(ON_START_DELAY_DURATION_MILLIS)
+      runCurrent()
+      assertThat(worker.onStartFinished).isTrue()
+      assertThat(dispatcher.dispatchCount).isEqualTo(1) // no new dispatch done
+      handle.unbind()
+      assertThat(worker.onStopThread!!.id).isEqualTo(Thread.currentThread().id)
+      assertThat(worker.onStopFinished).isTrue()
+    }
+  }
+}
+
+@OptIn(ExperimentalCoroutinesApi::class, InternalCoroutinesApi::class)
+private class ImmediateDispatcher(
+  scheduler: TestCoroutineScheduler,
+  private val delegate: TestDispatcher = StandardTestDispatcher(scheduler)
+) : CoroutineDispatcher(), Delay by delegate {
+  private var threadId: Long? = null
+  var dispatchCount = 0
+    private set
+
+  override fun isDispatchNeeded(context: CoroutineContext): Boolean {
+    val expectedThreadId = threadId ?: return true
+    return Thread.currentThread().id != expectedThreadId
+  }
+
+  override fun dispatch(context: CoroutineContext, block: Runnable) {
+    ++dispatchCount
+    delegate.dispatch(context, block)
+  }
+
+  fun setThreadId() {
+    threadId = Thread.currentThread().id
+  }
+}
+
+private class TestCoroutineWorker : CoroutineWorker {
+  var onStartStarted = false
+  var onStartFinished = false
+  var onStartThread: Thread? = null
+  var onStopCause: Throwable? = null
+  var onStopFinished = false
+  var onStopThread: Thread? = null
+
+  private var _doOnStart: suspend () -> Unit = {}
+
+  fun doOnStart(block: suspend () -> Unit) {
+    _doOnStart = block
+  }
+
+  override suspend fun onStart(scope: CoroutineScope) {
+    onStartStarted = true
+    onStartThread = Thread.currentThread()
+    try {
+      scope.launch { awaitCancellation() }
+      delay(ON_START_DELAY_DURATION_MILLIS)
```

Comment:
why this delay?

Reply:
I think I initially wanted to test something when the binding was in the state of started, but not yet finished (onStart suspended). Going over current tests, I don't think it is of any importance, so I'll just remove it. Thanks!